DRep expiry update with number of dormant epochs

aniketd opened the pull request:
# Description

<!-- Add your description here, if it fixes a particular issue please provide a
[link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=)
to the issue. -->

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

aniketd committed:
Fix DRep expiry updates based on numDormantEpochs

- Remove DRepPulser.dormantEpoch as it is no longer required.
- Add updateDormantDRepExpiry to only update active DRep expiries.
- Fix updateNumDormantEpochs to take current proposals, rather than old ones.
- In GOVCERT, make ConwayUpdateDRep calculate (epochNo + drepActivity - numDormantEpochs).
- Add relevant tests.

aniketd committed:
Update changelog

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs:
```diff
@@ -254,3 +254,24 @@ instance
+                if e < currentEpoch
+                  then e
+                  else binOpEpochNo (+) numDormantEpochs e
+             )
```
We should not update the expiry only if DRep has already expired. But the current value in the `drepExpiry` does not reflect the actual value. The actual value is `drepExpiry +  numDormantEpochs`
```suggestion
          %~ ( \curExpiry ->
                let newExpiry = binOpEpochNo (+) numDormantEpochs curExpiry
                if newExpiry < currentEpoch
                  then curExpiry
                  else newExpiry
             )
```

aniketd replied:
:bow: brilliant catch!

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/GovCert.hs:
```diff
@@ -207,7 +213,11 @@ conwayGovCertTransition = do
                 ( \drepState ->
                     drepState
-                      & drepExpiryL .~ addEpochInterval cgceCurrentEpoch ppDRepActivity
+                      & drepExpiryL
```
This scenario didn't even occur, it is very nice that you thought of this.

Can you also check that we do the same thing where we refresh the DRep expiry on a vote submission.

I can't seem to find where we do that, which might be a bug if it is missing.

But the idea is that it is possible scenario:
* there to be no proposals when the epoch starts and the number of dormant epochs is `numDormantEpochs > 0`
* someone submits a proposal
* a DRep votes on that proposal, thus refreshing its expiry `curEpoch + ppDRepActivity`
* epoch boundary comes we incorrectly update the expiry for that DRep to be further in the future than it has to be: `curEpoch + ppDRepActivity + numDormantEpochs`

In other words similar problem as with update certificate, except it needs more conditions to occur.

aniketd replied:
Awesome catch again :bow:

lehins reviewed (commented):
Haven't reviewed it thoroughly yet, but here are a couple of important comments

aniketd committed:
Fix dormant epochs calc at multiple places.

Add `vsActualDRepExpiry`.

Consider that effective DRep expiry is always
drepExpiry + numDormantEpochs at all places.

aniketd committed:
Make resetting numDormantEpochs part of udpate func

aniketd committed:
Refactor updating DRep expiry

aniketd committed:
Minor tweaks

aniketd committed:
Adjust the changelog

teodanciu commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/GovCert.hs:
```diff
@@ -204,14 +211,8 @@ conwayGovCertTransition = do
-                )
-                cred
-                vsDReps
+              updateDRepExpiry ppDRepActivity cgceCurrentEpoch (vState ^. vsNumDormantEpochsL) cred $
```
Is it necessary to have a separate `Map.adjust` in `updateDRepExpiry` , rather than having it return the EpochNo  as `binOpEpochNo (-) (addEpochInterval currentEpoch ppDRepActivity) numDormantEpochs`  and then setting it here with `drepExpiryL` ?

lehins replied:
Yeah, I also agree with you. Moreover, I think it will be better to turn `updateDRepExpiry` to handle just one DRepState, rather than also do the `adjust` in a Map. This way it will look like that:
```haskell
 Map.adjust
   (\drepState -> 
     updateDRepExpiry
       ppDRepActivity
       cgceCurrentEpoch
       (vState ^. vsNumDormantEpochsL)
       (drepState & drepAnchorL .~ mAnchor))
   cred
   vsDReps

teodanciu commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs:
```diff
@@ -254,3 +254,28 @@ instance
+  VState era ->
+  VState era
+updateDormantDRepExpiry currentEpoch numDormantEpochs vState =
+  if numDormantEpochs == EpochNo 0
```
We are already checking this condition before calling the function, so it gets invoked only if numDormantEpochs /= EpochNo 0 , no?  Is there a point to check both before and within the function?

teodanciu replied:
If we're checking inside the function, then maybe we don't need to pass the `numDormantEpochs` as a parameter, since it's already inside the VState?

lehins replied:
Very nice suggestion, I noticed the same thing

teodanciu commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs:
```diff
@@ -254,3 +254,28 @@ instance
+      drepExpiryL
+        %~ \currentExpiry ->
+          let actualExpiry = binOpEpochNo (+) numDormantEpochs currentExpiry
+           in if actualExpiry < currentEpoch
```
Just for my benefit, the difference from the previous implementation for the case in which we have proposals  and non-zero dormant epochs is here, right?  We don't extend the expiry if it's already expired, am I understanding correctly?

aniketd replied:
Yes, you are correct. :+1:

teodanciu commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/GovCert.hs:
```diff
@@ -227,3 +228,23 @@ conwayGovCertTransition = do
+      .~ binOpEpochNo
+        (-)
+        (addEpochInterval currentEpoch ppDRepActivity)
+        numDormantEpochs
```
Won't this overflow for big `numDormantEpochs` ?

lehins replied:
It would indeed overflow for large enough `numDormantEpochs`, except `numDormantEpochs` is incremented with each epoch and is not a user supplied value, so it has no chance of overflowing.

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs:
```diff
@@ -209,27 +210,26 @@ conwayCertsTransition = do
-                      & certVStateL . vsDRepsL %~ (<&> (drepExpiryL %~ binOpEpochNo (+) numDormantEpochs))
-                      & certVStateL . vsNumDormantEpochsL .~ EpochNo 0
+             in if hasProposals && numDormantEpochs /= EpochNo 0
+                  then certState & certVStateL %~ updateDormantDRepExpiry currentEpoch numDormantEpochs
```
Slight simplification:
`updateDormantDRepExpiry` can lookup `vsNumDormantEpochsL`, we don't need to extract it and pass it as an argument.
```suggestion
             in if hasProposals
                  then certState & certVStateL %~ updateDormantDRepExpiry currentEpoch
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs:
```diff
@@ -209,27 +210,26 @@ conwayCertsTransition = do
       -- This will execute in mutual-exclusion to the previous updates to DRep expiry,
       -- because if there are no proposals to vote on , there will be no votes either.
-      let drepActivity = pp ^. ppDRepActivityL
+      let numDormantEpochs1 = certState' ^. certVStateL . vsNumDormantEpochsL
```
Other binding with the same name will go away due to my suggestion.
```suggestion
      let numDormantEpochs = certState' ^. certVStateL . vsNumDormantEpochsL
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs:
```diff
@@ -209,27 +210,26 @@ conwayCertsTransition = do
-          certStateWithDRepExpiryUpdated = certState' & certVStateL . vsDRepsL .~ updatedVSDReps
+
+      -- Final CertState with updates to DRep expiry based on new proposals and votes on existing proposals
+      let certStateWithDRepExpiryUpdated = certState' & certVStateL . vsDRepsL .~ updatedVSDReps
```
Let's change this to a modification lens. this way we can avoid duplicating this above:` (certState' ^. certVStateL . vsDRepsL)`
```suggestion
      let certStateWithDRepExpiryUpdated = certState' & certVStateL . vsDRepsL %~ updateVSDReps
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:
```diff
@@ -1377,22 +1415,47 @@ submitConstitution prevGovId = do
+  HasCallStack =>
+  Credential 'DRepRole (EraCrypto era) ->
+  ImpTestM era ()
+expectDRepResigned drep = do
```
DReps do not resign, it is CC members that can resign. We should avoid conflicting terminology.
```suggestion
expectDRepNotRegistered ::
  HasCallStack =>
  Credential 'DRepRole (EraCrypto era) ->
  ImpTestM era ()
expectDRepNotRegistered drep = do
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:
```diff
@@ -362,6 +371,35 @@ registerDRep = do
+  submitTxAnn_ "Update DRep" $
+    mkBasicTx mkBasicTxBody
+      & bodyTxL . certsTxBodyL
+        .~ SSeq.singleton (UpdateDRepTxCert drep SNothing)
```
Let's add a bit of randomness in:
```suggestion
updateDRep drep = do
  mAnchor <- arbitrary
  submitTxAnn_ "Update DRep" $
    mkBasicTx mkBasicTxBody
      & bodyTxL . certsTxBodyL
        .~ SSeq.singleton (UpdateDRepTxCert drep mAnchor)
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/EpochSpec.hs:
```diff
@@ -146,59 +146,223 @@ dRepSpec =
       expectNumDormantEpochs 0
 
-      -- epoch 0
+      -- epoch 0: we submit a proposal, pulser is empty
```
Pulser isn't relevant for expiry
```suggestion
      -- epoch 0: we submit a proposal
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/EpochSpec.hs:
```diff
@@ -146,59 +146,223 @@ dRepSpec =
+          & ppDRepActivityL .~ EpochInterval 4
+      (drep1, _, _) <- setupSingleDRep 1_000_000 -- Receives an expiry update transaction certificate
+      (drep2, _, _) <- setupSingleDRep 1_000_000 -- Turns inactive due to natural expiry
+      (drep3, _, _) <- setupSingleDRep 1_000_000 -- Resigns and gets deleted
```
```suggestion
      (drep3, _, _) <- setupSingleDRep 1_000_000 -- Unregisters and gets deleted
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/EpochSpec.hs:
```diff
@@ -146,59 +146,223 @@ dRepSpec =
+      expectDRepExpiry drep1 9
+      expectDRepExpiry drep2 6
+
+      passEpoch -- entering epoch 6
```
Count is off in the comment. 
```suggestion
      passEpoch -- entering epoch 7
```

Maybe we can `passNEpochsChecking`, where it would be like `passNEpochs` except it would also accept an expectation that would be checked for every epoch being crossed. This way we could write:
```haskell
      passNEpochsChecking 3 $ do -- passing epoch 6 through 7
        expectCurrentProposals
        expectPulserProposals
        expectNumDormantEpochs 0
        expectDRepExpiry drep1 9
        expectDRepExpiry drep2 6
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs:
```diff
@@ -208,9 +208,9 @@ returnProposalDeposits removedProposals oldUMap =
-  if dormantEpoch pulser
+updateNumDormantEpochs :: Proposals era -> VState era -> VState era
+updateNumDormantEpochs ps vState =
+  if null $ ps ^. pPropsL
```
I think we need to remove the expired proposals, since they cannot be voted on.
```suggestion
  if null $ OMap.filter (\gas -> curEpochNo <= gasExpiresAfter gas) (ps ^. pPropsL)
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/EpochSpec.hs:
```diff
@@ -146,59 +146,223 @@ dRepSpec =
+      expectPulserProposals
+      expectNumDormantEpochs 0
+      expectDRepExpiry drep1 9
+      expectDRepExpiry drep2 6
```
That is an interesting case. Technically the proposal that is still in the system is expired, so no DRep can vote on it, but because there proposals set is not empty dormant epoch is not being incremented. 
I think we might need to do something about this edge case. I made a suggestion in the EPOCH rule about this
```suggestion
      -- there is still one proposal in the state, but it is expired:
      expectCurrentProposals
      expectPulserProposals
      expectNumDormantEpochs 1
      expectDRepExpiry drep1 9
      expectDRepExpiry drep2 6
```

lehins reviewed (commented):
Looks like there is still a small issue with dormant epochs counting, which I noticed thanks to tests in this PR. 
Other than that and a few minor suggestions this PR looks very nice

aniketd committed:
Address all review comments

aniketd committed:
Add voting to the DRep expiry update tests

aniketd committed:
Add passNEpochsChecking

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs:
```diff
@@ -254,3 +253,27 @@ instance
+    else
+      vState
+        & vsNumDormantEpochsL .~ EpochNo 0
+        & vsDRepsL %~ fmap updateExpiry
```
`Functor` for `Map` uses lazy implementation of `map`.
```suggestion
        & vsDRepsL %~ Map.map updateExpiry
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/CertState.hs:
```diff
@@ -628,6 +629,10 @@ vsCommitteeStateL = lens vsCommitteeState (\vs u -> vs {vsCommitteeState = u})
 
+vsActualDRepExpiry :: Credential 'DRepRole (EraCrypto era) -> VState era -> Maybe EpochNo
+vsActualDRepExpiry cred vs =
+  fmap (binOpEpochNo (+) (vsNumDormantEpochs vs) . drepExpiry) $ Map.lookup cred $ vsDReps vs
```
Slight simplification
```suggestion
  binOpEpochNo (+) (vsNumDormantEpochs vs) . drepExpiry <$> Map.lookup cred (vsDReps vs)
```

lehins reviewed (approved):
:+1: Nice. Thank you.

lehins committed:
Apply suggestions from code review
